Add snippets

andyw8 opened the pull request:
This PR adds a set of snippets to the add-on (see https://zed.dev/docs/snippets)

~They were added to Ruby LSP in https://github.com/Shopify/vscode-ruby-lsp/pull/150. I believe they were derived from [Textmate's Ruby bundle](https://github.com/textmate/ruby.tmbundle).~ (incorrect)

The `vscode-ruby-lsp` repo was later combined into the `ruby-lsp` repo, so they now live at https://github.com/Shopify/ruby-lsp/blob/main/vscode/snippets.json. The file I'm adding here is an exact copy that file.

I think there's a opportunity to improve the snippets, so consider this as a starting point.

~I'm currently investigating why snippets are appearing twice:~ solved by https://github.com/zed-industries/zed/pull/28940

Note: It's possible that snippet completion may be later moved into the LSP: https://github.com/Shopify/ruby-lsp/issues/1874

TODO:
- [x] Check if the `LICENSE` needs updated to mention that these were sourced from Ruby LSP.
- [x] Wait for https://github.com/zed-industries/zed/issues/15846

andyw8 commented:
Also opened an issue for the snippets being listed twice: https://github.com/zed-industries/zed/issues/28661

redforks commented:
I think I fixed it, https://github.com/zed-industries/zed/pull/28940

andyw8 commented:
@redforks thank you! I'll try on the next release.

andyw8 commented on LICENSE-APACHE:
```diff
@@ -220,3 +220,15 @@ Apache License
+
+---
+
+The contents of snippets.json are derived from Ruby LSP:
```
:wave: @vinistock, are you ok with this attribution?

vinistock replied:
Let's include the copyright line.

```
Copyright (c) 2022-present, Shopify Inc.
https://github.com/Shopify/ruby-lsp
  
  Released under the MIT license
  https://github.com/Shopify/ruby-lsp/blob/main/vscode/LICENSE.txt
```

joeldrapper commented:
I love the idea of having a nice set of default Ruby snippets, except the indentation is messed up and some of the shortcuts seem confusing to me. For example `def` here is a snippet for a singleton method, which seems odd.

andyw8 commented:
In https://github.com/Shopify/vscode-ruby-lsp/pull/150 there were snippets for both `def` and `def self.` but for some reason the current version only has `def self.` I've added it back in. I've also renamed the singleton one's prefix to `defs` (as it was in the [2006 original](https://github.com/textmate/ruby.tmbundle/blob/master/Snippets/def%20self%20..%20end%20%20(defs).plist)).

I think the indentation is the same issue as reported in https://github.com/zed-industries/zed/issues/15846

joeldrapper commented:
Looks like it. `defs` is nice! I’ve run into indentation issues with snippets before and couldn’t figure it out. Looks like this is why. It seems like the snippet templates were designed to assume the editor could always figure out the indentation by itself but I don’t know if that can always be the case. It would be useful if the template could store where to outdent and indent and then have it apply whatever your indentation characters are set to.

vinistock commented on LICENSE-APACHE:
```diff
@@ -220,3 +220,15 @@ Apache License
+https://github.com/Shopify/ruby-lsp
+    https://github.com/Shopify/ruby-lsp/blob/main/vscode/LICENSE.txt
+
+which are in turn derived from:
+
+https://github.com/textmate/ruby.tmbundle
+    https://github.com/textmate/ruby.tmbundle#license
```
This is not correct. It's the Text Mate grammars and language configuration files that are derived from a series of adaptions, not the snippets.

vinistock reviewed (commented):
Ideally, we would complete the work to move these to the Ruby LSP server to make it available for any editor without needing to copy the snippets around (in addition to making them smarter).

That said, if Zed wants a short term solution, this should do the trick.

vitallium commented:
Thanks for adding the snippets! Before merging this, I'd like to check with the Zed team first to make sure they’re okay with modifying the license file.

@osiewicz Hi, sorry for pinging you directly, but could you please check if modifying the license file to include copyrights for other components is OK from the Zed team's perspective? Thanks!

joeldrapper commented:
@vitallium can I also suggest we wait on https://github.com/zed-industries/zed/issues/15846 before merging this? The snippets look great, but I tried using them in a dev build locally and found I wanted to disable them almost immediately due to indentation issues.

vitallium commented:
> @vitallium can I also suggest we wait on [zed-industries/zed#15846](https://github.com/zed-industries/zed/issues/15846) before merging this? The snippets look great, but I tried using them in a dev build locally and found I wanted to disable them almost immediately due to indentation issues.

Yes, I played around with this PR a bit, and I also noticed some oddities with spacing and indentation.

osiewicz commented:
@vitallium Yep, these license changes look good to me (not a lawyer though).

jonyoi commented:
I tried these snippets and think they can be improved by adding default editable placeholders—editable with just one backspace. Sublime Text does this well for def and do blocks.

Below are example snippets I created for zed to show what I mean, you will notice I am using ClassName, method_name, and ModuleName as defaults, which also follow Ruby naming conventions, which will be nice reminder to users.

It would be great if Zed could either include similar snippets, allow editing of default ones, or prioritize custom ones. Currently, typing def shows a suggestion that doesn’t insert anything. I’m hoping once proper snippets are added, they’ll take effect, so I don’t have to name mine defm instead of just def.

I tried these snippets and think they can be improved by adding default editable placeholders—editable with just one backspace. Sublime Text does this well for def and do blocks.

Below are example snippets I created for zed to show what I mean, you will notice I am using ClassName, method_name, and ModuleName as defaults, which also follow Ruby naming conventions, which will be nice reminder to users.

It would be great if Zed could either include similar snippets, allow editing of default ones, or prioritize custom ones. Currently, typing def shows a suggestion that doesn’t insert anything. I’m hoping once proper snippets are added, they’ll take effect, so I don’t have to name mine defm instead of just def.

```
{
  "def": {
    "prefix": "def",
    "body": [
      "def ${1:method_name}",
      "  $0",
      "end"
    ],
    "description": "Ruby method definition"
  },
  "class": {
    "prefix": "class",
    "body": [
      "class ${1:ClassName}",
      "  $0",
      "end"
    ],
    "description": "Ruby class definition"
  },
  "module": {
    "prefix": "module",
    "body": [
      "module ${1:ModuleName}",
      "  $0",
      "end"
    ],
    "description": "Ruby module definition"
  },
  "do": {
    "prefix": "do",
    "body": [
      "do |${1:variable}|",
      "  $0",
      "end"
    ],
    "description": "Ruby do block definition"
  }
}
```

andyw8 commented:
This should be good to merge now that the indentation issue is resolved.
